### PR TITLE
Define leading '_' as API for creating temporaries

### PR DESCRIPTION
**Pull request by jackkoenig**
Note, I've rendered the updated docs [here](https://gist.github.com/jackkoenig/5777f77ba7c1a66c574d6f566bc31724).

Chisel and FIRRTL have long used signals with names beginning with an
underscore as an API to specify that the name does not really matter.
Tools like Verilator follow a similar convention and exclude signals
with underscore names from waveform dumps by default. With the
introduction of compiler-plugin prefixing in Chisel 3.4, the convention
remained but was hard for users to use unless the unnnamed signal
existed outside of any prefix domain. Notably, unnamed signals are most
useful when creating wires inside of utility methods which almost always
results in the signal ending up with a prefix.

With this commit, Chisel explicitly recognizes signals whos val names
start with an underscore and preserve that underscore regardless of any
prefixing. Chisel will also ignore such underscores when generating
prefixes based on the temporary signal, preventing accidental double
underscores in the names of signals that are prefixed by the temporary.

Here's a specific example of the change this makes, Given:
```scala
class Example extends Module {
  val foo, bar = IO(Input(UInt(8.W)))
  val out = {
    val port = IO(Output(UInt(8.W)))
    val _sum = foo + bar
    val _sum2 = Wire(UInt(8.W))
    _sum2 := foo + 1.U
    port := _sum + _sum2
    port
  }
}
```
Chisel master (and 3.4.3) currently compile this to:
```
module Example :
  input clock : Clock
  input reset : UInt<1>
  input foo : UInt<8>
  input bar : UInt<8>
  output out : UInt<8>

  node _out__sum_T = add(foo, bar)
  node out__sum = tail(_out__sum_T, 1)
  wire out__sum2 : UInt<8>
  node _out__sum2_T = add(foo, UInt<1>("h1"))
  node _out__sum2_T_1 = tail(_out__sum2_T, 1)
  out__sum2 <= _out__sum2_T_1
  node _out_port_T = add(out__sum, out__sum2)
  node _out_port_T_1 = tail(_out_port_T, 1)
  out <= _out_port_T_1
```

Now this is compiled to:
```
  module Example :
    input clock : Clock
    input reset : UInt<1>
    input foo : UInt<8>
    input bar : UInt<8>
    output out : UInt<8>

    node _out_sum_T = add(foo, bar)
    node _out_sum = tail(_out_sum_T, 1)
    wire _out_sum2 : UInt<8>
    node _out_sum2_T = add(foo, UInt<1>("h1"))
    node _out_sum2_T_1 = tail(_out_sum2_T, 1)
    _out_sum2 <= _out_sum2_T_1
    node _out_port_T = add(_out_sum, _out_sum2)
    node _out_port_T_1 = tail(_out_port_T, 1)
    out <= _out_port_T_1 
```

Note that `out__sum` and `out__sum2` become `_out_sum` and `_out_sum2`. This change comes from Chisel effectively recognizing the leading `_` is not actually part of the name, but is the signifier of a temporary so it is stripped from the "seed" and moved to the front of the final name.

Also note that `_out__sum2_T` becomes `_out_sum2_T`. This functionality looks similar to the previous paragraph, but is actually a totally different change where Chisel, again recognizing that `_` is the signifier of a temporary and not really part of the name, actually ignores the leading `_` for generating prefixes from signals.

Between these two changes, I think the naming behavior is a pretty nice improvement over the current behavior.

### Contributor Checklist

- [x] Did you add Scaladoc to every public function/method?
- [x] Did you add at least one test demonstrating the PR?
- [x] Did you delete any extraneous printlns/debugging code?
- [x] Did you specify the type of improvement?
- [x] Did you add appropriate documentation in `docs/src`?
- [x] Did you state the API impact?
- [x] Did you specify the code generation impact?
- [x] Did you request a desired merge strategy?
- [x] Did you add text to be included in the Release Notes for this change?

#### Type of Improvement

 - new feature/API    

#### API Impact

This adds a new API to let users creating "temporary" signals. I will note that this might have a s

#### Backend Code Generation Impact

This could potentially rejigger a lot of names. On the one hand, I see this as a bug fix (and thus backportable) because I think the prior behavior of ignoring users defining signals with leading `_` as sort of a bug. On the other hand, this has the potential to cause a big change in naming when bumping and users may not appreciate that on a minor release.

#### Desired Merge Strategy

  - Squash

#### Release Notes

Define leading '_' as API for creating temporaries. Chisel will now preserve the leading `_` for any val-named signal which signifies to FIRRTL and other backend tools that the name does not matter. Chisel also now ignores the leading `_` in prefixes generated from such "unnamed signals" which reduces the number of cases of double `__` in the middle of signal names.

### Reviewer Checklist (only modified by reviewer)
- [ ] Did you add the appropriate labels?
- [ ] Did you mark the proper milestone (Bug fix: `3.4.x`, [small] API extension: `3.5.x`, API modification or big change: `3.6.0`)?
- [ ] Did you review?
- [ ] Did you check whether all relevant Contributor checkboxes have been checked?
- [ ] Did you do one of the following when ready to merge:
  - [ ] Squash: You/ the contributor `Enable auto-merge (squash)`, clean up the commit message, and label with `Please Merge`.
  - [ ] Merge: Ensure that contributor has cleaned up their commit history, then merge with `Create a merge commit`.

### CODE REVIEW

**Comment by jackkoenig** [comment]
I've marked this 3.6 because it has a fairly large impact on code gen stability when bumping and I am not sure that's a great thing to do in a minor release. That being said I could be convinced that this makes sense in 3.5.4.

File: core/src/main/scala/chisel3/RawModule.scala
Comment (jackkoenig) [inline core/src/main/scala/chisel3/RawModule.scala:69]: I realized that my changes here make the "defaultPrefix" argument of `_computeName` redundant so I need to further clean that up.

File: core/src/main/scala/chisel3/internal/Builder.scala
Comment (jackkoenig) [inline core/src/main/scala/chisel3/internal/Builder.scala:189]: This code can get a lot simpler now, TODO fix.

File: docs/src/explanations/naming.md
Comment (mwachs5) [inline docs/src/explanations/naming.md:206]: nice. I was coming here to say "doesn't this deserve a docs update" and here it is!

File: src/test/scala/chiselTests/naming/PrefixSpec.scala
Comment (mwachs5) [inline src/test/scala/chiselTests/naming/PrefixSpec.scala:471]: should we show that adding a `suggestName` overrides the `_` behavior, whatever its?

**Comment by seldridge** [comment]
> Define leading '_' as API for creating temporaries. Chisel will now preserve the leading _ for any val-named signal which signifies to FIRRTL and other backend tools that the name does not matter. Chisel also now ignores the leading _ in prefixes generated from such "unnamed signals" which reduces the number of cases of double __ in the middle of signal names.

This release notes info is great.

WDYT about codifying some of this into a Chisel spec?

Possibly useful, but the way this is evolving on the FIRRTL Dialect side is that ops have a _name kind_ enum that qualify the utility of keeping a name around and let the compiler reason about this for different optimization levels. Currently there are two values: (1) interesting and (2) droppable. This matches the existing non-underscore (droppable) and named (interesting) convention, but allows for further enumeration going forward and avoids permanently encoding this information in the leading character of a string.

File: src/test/scala/chiselTests/naming/PrefixSpec.scala
Comment (jackkoenig) [reply src/test/scala/chiselTests/naming/PrefixSpec.scala:471]: I'm documenting how `suggestName` affects prefixing

File: docs/src/explanations/naming.md
Comment (mwachs5) [inline docs/src/explanations/naming.md:252]: but what happens if I `suggestName("mySuggestedName")` on `diff`? Is it `out_mySuggestedName` or just `mySuggestedName`?

TBH I lost track of what it actually does, then it's a question is that a bug or a feature

File: src/test/scala/chiselTests/naming/PrefixSpec.scala
Comment (jackkoenig) [reply src/test/scala/chiselTests/naming/PrefixSpec.scala:499]: Done, see the additional docs (rendered [here](https://gist.github.com/jackkoenig/5777f77ba7c1a66c574d6f566bc31724))

File: docs/src/explanations/naming.md
Comment (jackkoenig) [reply docs/src/explanations/naming.md:251]: This is [already documented](https://www.chisel-lang.org/chisel3/docs/explanations/naming.html#suggest-a-signals-name-or-the-instance-name-of-a-module)

> If you want to specify the name of a signal, you can always use the .suggestName API. Please note that the suggested name will still be prefixed (including by the plugin). You can always use the noPrefix object to strip this.

File: docs/src/explanations/naming.md
Comment (jackkoenig) [reply docs/src/explanations/naming.md:228]: The example is right below this line 😉, pasting from the mdoc-output of this section:

-------------------

Note that using `.suggestName` does *not* affect prefixes derived from val names; however, it *can* affect prefixes derived from connections (eg. `:=)`:
```scala
class ConnectionPrefixExample extends Module {
  val in0 = IO(Input(UInt(2.W)))
  val in1 = IO(Input(UInt(2.W)))

  val out0 = IO(Output(UInt()))
  val out1 = IO(Output(UInt()))

  out0 := {
    // out0_sum
    val sum = in0 + in1
    sum + 1.U
  }
  // Comes after so does *not* affect prefix above
  out0.suggestName("foo")

  // Comes before so *does* affect prefix below
  out1.suggestName("bar")
  out1 := {
    // bar_diff
    val diff = in0 - in1
    diff + 1.U
  }
}
```
```verilog
module ConnectionPrefixExample(
  input        clock,
  input        reset,
  input  [1:0] in0,
  input  [1:0] in1,
  output [1:0] foo,
  output [1:0] bar
);
  wire [1:0] out0_sum = in0 + in1; // @[naming.md 235:19]
  wire [1:0] bar_diff = in0 - in1; // @[naming.md 245:20]
  assign foo = out0_sum + 2'h1; // @[naming.md 236:9]
  assign bar = bar_diff + 2'h1; // @[naming.md 246:10]
endmodule
```

Comment (mwachs5) [reply docs/src/explanations/naming.md:228]: but neither of these examples is showing the suggestName on `diff` itself

Comment (mwachs5) [reply docs/src/explanations/naming.md:228]: but i think you covered that with the link to the "already covered" case about suggestName

Comment (mwachs5) [reply docs/src/explanations/naming.md:228]: Wait sorry my response was to the wrong point, be back in a sec

Comment (mwachs5) [reply docs/src/explanations/naming.md:228]: just looking for an example to demonstrate what you are saying in " Note that using `.suggestName` does *not* affect prefixes derived from val names"

Comment (jackkoenig) [reply docs/src/explanations/naming.md:228]: I've added another example to that case, here's the updated text:

----------------------

Note that using `.suggestName` does **not** affect prefixes derived from val names;
however, it _can_ affect prefixes derived from connections (eg. `:=`):

```scala
class ConnectionPrefixExample extends Module {
  val in0 = IO(Input(UInt(2.W)))
  val in1 = IO(Input(UInt(2.W)))

  val out0 = {
    val port = IO(Output(UInt()))
    // out0_mul
    val mul = in0 * in1
    port := mul + 1.U
    port
  }
  // The prefix was derived from "out0" above, suggestName does not affect prefix
  out0.suggestName("foo")

  val out1 = IO(Output(UInt()))
  val out2 = IO(Output(UInt()))


  out1 := {
    // out1_sum
    val sum = in0 + in1
    sum + 1.U
  }
  // Comes after so does *not* affect prefix above
  out1.suggestName("bar")

  // Comes before so *does* affect prefix below
  out2.suggestName("fizz")
  out2 := {
    // fizz_diff
    val diff = in0 - in1
    diff + 1.U
  }
}
```
```verilog
module ConnectionPrefixExample(
  input        clock,
  input        reset,
  input  [1:0] in0,
  input  [1:0] in1,
  output [3:0] foo,
  output [1:0] bar,
  output [1:0] fizz
);
  wire [3:0] out0_mul = in0 * in1; // @[naming.md 232:19]
  wire [1:0] out1_sum = in0 + in1; // @[naming.md 245:19]
  wire [1:0] fizz_diff = in0 - in1; // @[naming.md 255:20]
  assign foo = out0_mul + 4'h1; // @[naming.md 233:17]
  assign bar = out1_sum + 2'h1; // @[naming.md 246:9]
  assign fizz = fizz_diff + 2'h1; // @[naming.md 256:10]
endmodule
```

Comment (jackkoenig) [reply docs/src/explanations/naming.md:228]: Actually I'm going to tweak it to make it even more obvious by putting the `.suggestName` on `port`.

Comment (jackkoenig) [reply docs/src/explanations/naming.md:228]: Okay further updated

**Comment by jackkoenig** [comment]
> WDYT about codifying some of this into a Chisel spec?

@seldridge I think that would be a good thing to do. It would be fairly straightforward to specify the semantics/rules for emitting FIRRTL at least. It's a bit more difficult for defining down to Verilog until we have stronger guarantees on how things lower from FIRRTL to Verilog but perhaps it's time to at least get the ball rolling.

**Comment by jackkoenig** [comment]
@mwachs5 I think this is ready to :shipit:

**Comment by jackkoenig** [comment]
Note this is 3.6.0 though, might be nice to backport some of the docs changes to 3.5 so it shows up on the website

File: docs/src/explanations/naming.md
Comment (mwachs5) [inline docs/src/explanations/naming.md:84]: this is very helpful in an explanation doc!

**Review by mwachs5 — APPROVED** [review]
This is all crazy complicated but happy to approve the better docs of the status quo if we are open to changing them!

**Comment by mwachs5** [comment]
> 
Note this is 3.6.0 though, might be nice to backport some of the docs changes to 3.5 so it shows up on the website

I agree -- if we can backport the docs and tests without the `_` change that would be good

**Comment by jackkoenig** [comment]
Before we merge this, let me check if this actually impacts names that much. If the impact is very minor, we can backport the full PR. If the impact is huge, then I'll just backport the docs improvements.

**Comment by jackkoenig** [comment]
Okay this actually has zero impact on naming unless you intentionally name things with leading `_`. And since apparently that is not used at all in SiFive internal codebases at least, it seems that this is a reasonable thing to include in a minor release.

**Pull request merged**
